### DISCUSSION TITLE
Using Vitest to test async setup components with auto-imports?

### DISCUSSION THREAD

**Discussion by jshimkoski** [nuxt/nuxt#16879] [Questions]
Take this simple, ridiculous example component code:

```vue
<template>
  <div>{{ resultOfSomething }}: {{ test }}</div>
</template>

<script setup>
const { $doSomething } = useNuxtApp()
const { data: test } = await useFetch('/api/test/endpoint')

const resultOfSomething = $doSomething()
</script>
```

Let's write a basic "snapshot" test using Vitest and Vue-Test-Utils:

```javascript
import { describe, it, expect } from 'vitest'
import { shallowMount } from '@vue/test-utils'
import Component from '@/components/Component.vue'

describe('Component', () => {
  it('must match snapshot', () => {
    const wrapper = shallowMount(Component)
    expect(wrapper.element).toMatchSnapshot()
  })
})
```

You get errors stating that you can't test a Component with an `async setup` without having the component wrapped in `<Suspense>`.

So, you wrap it in `<Suspense>`:

```javascript
import { describe, it, expect } from 'vitest'
import { mount } from '@vue/test-utils'
import Component from '@/components/Component.vue'
import { Suspense } from 'vue'

const Parent = {
  components: {
    Suspense,
    Component
  },
  template: `
    <Suspense>
      <Component />
    </Suspense>
  `
}

describe('Component', () => {
  it('must match snapshot', () => {
    const wrapper = mount(Parent)
    expect(wrapper.element).toMatchSnapshot()
  })
})
```

The `<Suspense>` error is now gone, however, now you get dinged with `ReferenceError: useNuxtApp is not defined` because Vitest hit the Nuxt 3 auto-imports in your Component and it doesn't know how to handle them.

I have a few questions:

* How do you use Nuxt 3's auto-imports with Vitest?
* Is there a better way to handle the `Suspense` issue?
* Is there a better approach to all of this?

Thanks!

**Comment by lucavenir** [comment]
I have this exact same problem, but with simple composables logic testing. I think this might be "issuable" as a feature request or a bug.

Take the following trivial example (`composables/useCounter.ts`)
```ts
import { defineStore } from 'pinia';

export const useCounter = defineStore('counter', {
  state: () => ({ counter: 0 }),
  getters: {
    doubled: (state) => state.counter * 2,
  },
  actions: {
    reset() {
      this.counter = 0;
    },
  },
});
```

Then, I'd like to setup this test with Vitest. `composables/useCounter.test.ts`:
```ts
import { setActivePinia } from 'pinia';
import { beforeEach, describe, expect, it } from 'vitest';

describe('testing-store', () => {
  beforeEach(() => {
    setActivePinia(usePinia());
  });

  it('shall pass', () => {
    const c = useCounter();
    expect(c.counter).toBe(5);
  });
});
```

Both `usePinia` and `useCounter` aren't detected when performing tests (`vitest`), even though my IDE (VS Code) show no import errors at all! Therefore, I am unable to test composables with vitest at the moment.

**Comment by mitraecp** [comment]
i need help with autoimports in vitest too...

**Comment by anaroblesd** [comment]
Hello, I found a solution for this problem. I hope it works for you too. 
https://medium.com/@ana.robles.developer/nuxt-3-vitest-vue-test-utils-mockear-en-los-test-funciones-auto-importadas-2242297eb06d
